### ISSUE TITLE
install juicer (get quote for GPU node)

### ISSUE THREAD

**Issue by hurleyLi** [frazer-lab/cluster#153]
Hi @tatarsky ,

I wanted to install a software called juicer:
https://github.com/theaidenlab/juicer
One of the key command I want to use is HiCCUPS, which requires the installation of GNU CoreUtils and CUDA. Is this something we can implement on our cluster? I took a look at the installation site, and it's far beyond my understanding. Could you please take a look at the website and see if you can get it installed on the cluster. 

Thanks!
Hurley

**Comment by tatarsky** [comment]
Please consult with the local sysadmins for this. I don't have time for yet another new package.

**Comment by tatarsky** [comment]
Also as per the readme "You must have an NVIDIA GPU to install CUDA". There are no GPUs in this cluster. While some CUDA library calls "work" without them they usually not the more interesting calls that use GPU cores to process data. So this is unlikely to work.

**Comment by hurleyLi** [comment]
I see. Thanks

**Comment by tatarsky** [comment]
GPUs could be purchased but they require larger nodes (with PCI-E slots and spacing capable of holding them). Our nodes are very compact and dense. Usually you purchase special GPU holding nodes with the proper cooling and space.

**Comment by hurleyLi** [comment]
I guess SDSC may already have GPU nodes? maybe I should ask them and try to figure out how I can run my jobs on there

**Comment by tatarsky** [comment]
You could certainly try. Alternately ask Kelly for investigation into GPU capable nodes. Its not overly complex, but from my two second reading of that Git page (which is all I have right now for a bit) you won't get GPU acceleration of that process without GPUs ;)

**Comment by tatarsky** [comment]
I am taking a moment to look at juicer. No guarantees but my load dropped off a bit.

**Comment by tatarsky** [comment]
So basically I have no idea what the install process for this is. It may be self contained as its Java based. 

I'm going to place it here for now for you to try to gain that understanding with it. No module at this time because frankly not sure how to use it `/frazer01/software/juicer-0.9`

Coreutils is already installed as an RPM. Hopefully new enough.
Java 1.7 is already installed.

CUDA as noted really requires a GPU to do something. So this is a test without really that option.

The top level contains basically three "environments" of which the closest to ours is the "UGER" directory (UGER means apparently Univa's Gridengine which is the commercial version of SGE).

Try to cd into that area and see if you can make heads or tails of this package. 

```
./juicer.sh 
***! Reference sequence /broad/aidenlab/references/Homo_sapiens_assembly19.fasta does not exist
```

I assume you have to edit that path to point to your fasta files. 

I will look at this at a lower priority. I do not consider this my role to debug to this level but will answer questions if I can.

**Comment by hurleyLi** [comment]
The only tool and command I want to use in the java package is 

```
java -jar /frazer01/software/juicer-0.9/UGER/scripts/juicebox_tools.jar /projects/CARDIPS/pipeline/Hi-C/merged/CM/CM.merged_nodups.filtered.inter.intra_longRange.hic output.o
```

based on the manual, it requires NVIDIA GPU and CUDA. 
Kelly is open to install a GPU node, since we would pay SDSC anyway if we run the jobs there. Just wondering how much would it cost approximately, and how long would it take to have it installed. 

Thanks,

**Comment by tatarsky** [comment]
We'd have to talk a bit to clearly define what you mean by GPU node. 

Most have several GPUS and the type and performance of the GPU varies considerably. 

Most spots I deal with use commodity GPU cards that are really designed for graphics card usages (aka a monitor). Those run around $300-$1000 a card depending on the type of card and GPU memory you need. Beyond that you get into dedicated GPU cards of the Tesla line and they are considerably more expensive.

Do you have an idea what class of GPU card(s) you are wanting? You then build the node around the space/PCI bus requirements for that card. 

Also, if you have a UNIX desktop with say a fairly nice Nvidia graphics card you could actually attempt to get it working there first as a test.

**Comment by tatarsky** [comment]
To help a little more on those questions: see if you see references by other users of that software to 

-Number of GPU cores
-Level of GPU memory needed (on the card...not the node)

And then decide how many cards you need. Most entry level GPU nodes hold between 2 or 4 depending on the size of the card.

**Comment by hurleyLi** [comment]
I'm not sure what type of GPU is required, but I think in this case the GPU is used for processing the jobs, rather to show them on the monitor. 
Testing on my mac is good idea, just not sure what type of GPU card does my mac have and whether it supports NVIDIA GPU. let me dig a little more

**Comment by tatarsky** [comment]
What I'm getting at is many GPU nodes use consumer grade GPU cards that are REALLY made to show graphics on a monitor.  So they have the external connections for the monitor which sometimes get in the way of some chassis (node) to deploy them. Particularly more than one. And in particular the cooling becomes very important with more than one. 

I know you don't display anything on the monitor ;)

But the path you pick alters the chassis type of a "GPU node" a bit so its good to know what class of card you are thinking. 

Example common card I'm seeing many people use at the moment picked randomly but so you can see what I mean a bit. 

http://www.newegg.com/Product/Product.aspx?Item=N82E16814126101&cm_re=gtx_1080-_-14-126-101-_-Product

Compare to an entry level Tesla. No monitor connectors. Very expensive.

http://www.newegg.com/Product/Product.aspx?Item=N82E16814132041&cm_re=tesla-_-14-132-041-_-Product

As for the Mac side, see if you have a command like `nvidia-smi`.

**Comment by hurleyLi** [comment]
Based on the tool manual, both integrated GPU and dedicated GPU will work, but a dedicated GPU will work much more efficiently. They say "For best performance, use a dedicated GPU." 
This program does not have options to specify how many, I guess it automatically detect how many GPU cores are available? 
I do find in a paper they said "... on a workstation containing 4 NVIDIA GPUs. Using an
NVIDIA Tesla C2075 GPU (which contains 448 cores and is capable of highly parallelized computation), we obtain a 200- fold speedup over the CPU implementation. "
I guess that's very expensive ...

**Comment by tatarsky** [comment]
No argument on dedicated and on a "GPU node" that is implied by the adding of the card to a headless system. 

Well, that gives us a ballpark idea. If they get 200X from a 448 core Tesla we can perhaps get a quote for a single GTX1080 or something and see what sort of speed ups the lower cost route goes.

The key is where we put that card. I wonder if the old nodes have room and the proper power connectors. 

I can also ask for a simple quote for a simple GPU node.

**Comment by tatarsky** [comment]
That Tesla card BTW is an older model (back from 2011 actually) and no longer really what you would buy. So the 448 core is a better metric. Thats pretty common in commodity cards. The 1080 card I sent has 2560. So thats very important to understand as it was $700.

**Comment by tatarsky** [comment]
Also, is there a simple test case _I_ could do of this if I have a GPU system handy? Aka a packaged up set of some data and that juicer command?

**Comment by hurleyLi** [comment]
yes a simple command would be (command input output)

```
java -jar /frazer01/software/juicer-0.9/UGER/scripts/juicebox_tools.jar /projects/CARDIPS/pipeline/Hi-C/merged/CM/CM.merged_nodups.filtered.inter.intra_longRange.hic output.o
```

The input file (.hic) is really big, but maybe I can make a .hic file just for one chromosme to make it smaller for you to test

**Comment by tatarsky** [comment]
Tis OK. 40GB is doable. I'll see what I can do to locate a place to validate this path.

**Comment by hurleyLi** [comment]
Thanks Paul!

**Comment by hurleyLi** [comment]
I forgot to put the tool name in the command ...

```
java -jar juicebox_tools.jar hiccups input.hic output.o
```

Do you think we can put the graphics card on one of our long.queue nodes? If the conditions do not allow, e.g. power, cooling, etc, could you please get a quote on a GPU node. (I assume this GPU node can also be used as a normal node, right?)

**Comment by hurleyLi** [comment]
and a GPU node will use a designated GPU, not integrated GPU, is that right?

**Comment by tatarsky** [comment]
We cannot put a graphics card in the nodes we have now. They are very compact nodes. 

I've asked Kelly via email about getting a quote. Her approval for such items is required before I spend time on it. 

A GPU node can also be used as a normal node. And your designated GPU v.s. integrated comment I believe the answer is yes. If you have a precise definition of "integrated" I am assuming you mean not a motherboard intergrated GPU. These would be PCI-E cards.

**Comment by tatarsky** [comment]
No email response yet but kicked an email over to AHPC to at least get some ballparks.

**Comment by tatarsky** [comment]
Hmm. Well with java 8 which is all I have available on this remote node:

```
java -jar juicebox_tools.jar hiccups foo.hic output.o
No configurations specified, using default settings
Reading file: foo.hic
HiC file version: 8
Exception in thread "main" java.lang.NullPointerException
        at juicebox.tools.clt.juicer.HiCCUPS.run(HiCCUPS.java:250)
        at juicebox.tools.HiCTools.main(HiCTools.java:79)
```

foo.hic is your 60GB file. ideas?

**Comment by hurleyLi** [comment]
It looks like, from the google forum of this tool, it's the problem of old version juicer. Please find the latest version from here:
http://aidenlab.org/commandlinetools/
There seems to be two versions based on the CUDA version...

**Comment by tatarsky** [comment]
Downloading the 7.5 CUDA version. (As that matches the node I am trying this for you on)

**Comment by tatarsky** [comment]
Well got a little further but this node may be missing part of the CUDA software. Nobody uses CUDA from Java on this setup. So I'll look around.

```
java -jar juicebox_tools.7.5.jar hiccups foo.hic output.o
Sep 29, 2016 2:37:32 PM java.util.prefs.FileSystemPreferences$1 run
INFO: Created user preferences directory.
Reading file: foo.hic
HiC file version: 8
No valid configurations specified, using default settings
Unable to assess map sparsity; continuing with HiCCUPS
Default settings for 5kb, 10kb, and 25kb being used
Running HiCCUPS for resolution 5000
Exception in thread "main" java.lang.UnsatisfiedLinkError: Could not load the native library.
Error while loading native library "JCudaDriver-linux-x86_64" with base name "JCudaDriver"
java.lang.NullPointerException: No resource found with name '/lib/libJCudaDriver-linux-x86_64.so'
```

**Comment by hurleyLi** [comment]
Hi Paul, Kelly's approved to get us a quote on the GPU node. 
@kafrazer @nariai @hirokomatsui

**Comment by tatarsky** [comment]
OK. Sent to AHPC

**Comment by tatarsky** [comment]
Figured out where that lib comes from. They have it in the AWS tree but not the UGER one.
Appears to be doing something now.

**Comment by tatarsky** [comment]
So generic question....how long do you think this takes?

```
java -Djava.library.path=lib64/ -jar juicebox_tools.7.5.jar hiccups foo.hic output.o
Reading file: foo.hic
HiC file version: 8
No valid configurations specified, using default settings
Unable to assess map sparsity; continuing with HiCCUPS
Default settings for 5kb, 10kb, and 25kb being used
Running HiCCUPS for resolution 5000
(running...can see it on the GPU card)
```

**Comment by hurleyLi** [comment]
em... really have no idea how long it'll take... I've never run it. but it looks like it works. By specify`-c chr22`, you can run it only for one chromosome, which I assumes will take less time

**Comment by tatarsky** [comment]
OK ;)

I'll try that. It kicked out "0%" and I can't over occupy this node....

**Comment by tatarsky** [comment]
Much better....50% kicked out after a few seconds.

**Comment by tatarsky** [comment]
So takes about 40 seconds for this command:

```
java -Djava.library.path=lib64/ -jar juicebox_tools.7.5.jar hiccups -c chr22  foo.hic output.o
Reading file: foo.hic
HiC file version: 8
No valid configurations specified, using default settings
Unable to assess map sparsity; continuing with HiCCUPS
Default settings for 5kb, 10kb, and 25kb being used
Running HiCCUPS for resolution 5000
50% 
100% 
Running HiCCUPS for resolution 10000
50% 
100% 
Running HiCCUPS for resolution 25000
50% 
100% 
148 loops written to file:/home/paul/src/juicer-0.9/AWS/scripts/output.o/merged_loops
HiCCUPS complete
```

I will put the output.o somewhere you can see if its actually what you want. 

But this is a Titan X GPU card.  GeForce GTX TITAN. Now obsoleted by the GTX 1080 series. It has I believe 6GB of RAM and 2688 cores and was around $800 when it was new. Used they are < $500.

**Comment by tatarsky** [comment]
See if the contents of `~paul/output.o` make sense to you.

**Comment by hurleyLi** [comment]
it looks the right output! So just to make sure, if we purchase a GPU node, do we need to purchase a new node, and put the GPU card on top of it? And how many such GPU card do we need?

**Comment by tatarsky** [comment]
So "node" = Server or PC that can hold GPU cards. In our case something that goes in the rack.

GPU cards are PCI-E cards that require usually what are called "dual slot full width" bays. Which make them "large" in terms of the node chassis required to accommodate them.

They also require TWO particular types of power connectors each as they are real power consumers.

I have asked AHPC for a new Node capable of holding at least TWO GTX-1080 (the latest ones I'm seeing people use out there without buying Teslas). When I get a response I'll detail what that price is.

**Comment by tatarsky** [comment]
Still waiting on quote. Will call after a bit...

**Comment by tatarsky** [comment]
Emailed starter quote. Some additional items needed on it but its a start to give you a ballpark.

**Comment by tatarsky** [comment]
This was requested via email to be cancelled. Re-open if that changes.
